### PR TITLE
ability to sort and filter on actions

### LINKED ISSUE #151: Add feature to sort and/or filter by actions to recipe listing
Copied from: https://github.com/mozilla/normandy/issues/1040

> Being able to filter or sort by action type in the recipe listing can help quickly find the type of recipe that one cares about. It also stands out as the only default column that can't be sorted or filtered.

### PR DESCRIPTION
**Pull request by peterbe**
Fixes #151

### CODE REVIEW

**Comment by peterbe** [comment]
If you're wondering, I rebased this after I realized that my fancy new yarn 1.10 had added integrity values to all packages in the `yarn.lock`.

File: src/workflows/recipes/pages/RecipeListingPage.js [hunk 1]
```diff
@@ -28,9 +28,11 @@ import {
   getQueryParam,
   getQueryParamAsInt,
 } from 'console/state/router/selectors';
+import { getAllActions } from 'console/state/actions/selectors';
 
 @connect(
   (state, props) => ({
+    actions: getAllActions(state, new Map()),
```
Comment (peterbe) [inline src/workflows/recipes/pages/RecipeListingPage.js]: Can/should this be `actions: getAllActions(state) || new Map()` instead?

**Comment by peterbe** [comment]
screen shot 2018-10-08 at 12 38 35 pm

There *might* be a way to make that drop-down be a multi-value but I don't know how and I'm just not sure if it's worth it. If I tried to make it an array instead the XHR requests would become `/api/v2/recipe/?ordering=-last_updated&action=console-log,preference-experiment&page=1` (not the comma-separated list of action names)

File: src/workflows/recipes/pages/RecipeListingPage.js [hunk 2]
```diff
@@ -178,12 +105,100 @@ class RecipeListingPage extends React.PureComponent {
...
-    const { columns, count, ordering, pageNumber, recipes, status } = this.props;
+    const { columns, count, ordering, pageNumber, recipes, status, action, actions } = this.props;
+    const actionNames = Object.values(actions.toJS()).map(action => action.name);
+    actionNames.sort();
 
     const filters = this.getFilters();
 
     const filterIds = Object.keys(filters).map(key => `${key}-${filters[key]}`);
     const requestId = `fetch-filtered-recipes-page-${pageNumber}-${filterIds.join('-')}`;
+
```
Comment (peterbe) [inline src/workflows/recipes/pages/RecipeListingPage.js]: The change here is that I moved it from being a static object of functions to instead being defined *inside* the `render` method. I did that so I can dynamically (using closures) get the list of all possible action names. And I also, changed the function for `action`. The rest is just code moves.

**Comment by peterbe** [comment]
@rehandalal @mythmon Can you help me out here. Look at https://circleci.com/gh/mozilla/delivery-console/3531

The last line of the output says: `Too long with no output (exceeded 10m0s)` and I can't open the logs in raw. I'm going to try to ssh in to run it locally to see what's going on. 

Also, why are some unit tests taking 45 seconds to run? Are we doing too much concurrent test runs that it's CPU thrashing?

screen shot 2018-10-08 at 1 37 46 pm

**Comment by peterbe** [comment]
Ah! One clue is that we're running out of memory:
screen shot 2018-10-08 at 1 44 04 pm

Oh jest!

I'm going to see if we can run the tests in serial or something like that.

**Comment by peterbe** [comment]
The number of CPUs becomes uncontrollably big which isn't right when you use [`circleci/python:3.6-node-browsers`](https://github.com/mozilla/delivery-console/blob/377bc2f5d906da8d0a245532601217951f0326a5/.circleci/config.yml#L11)

See:

```bash
> circleci@9e4c489cf76b:~/repo$ node
> var os = require('os')
undefined
> os.cpus().length
36
```

When I, arbitrarily, pick max 2 workers things work better:

```bash
> circleci@9e4c489cf76b:~/repo$ time yarn run test:ci --maxWorkers 2
yarn run v1.9.4
$ CI=true yarn test --maxWorkers 2
$ node scripts/test.js test --env=jsdom --maxWorkers 2
...
 PASS  src/tests/components/data/QueryExtension.test.js
 PASS  src/tests/components/data/QueryExtensionListingColumns.test.js
 PASS  src/tests/components/data/QueryRecipeListingColumns.test.js
 PASS  src/tests/components/forms/FormActions.test.js
 PASS  src/tests/utils/router.test.js
 PASS  src/tests/utils/api.test.js

Test Suites: 68 passed, 68 total
Tests:       236 passed, 236 total
Snapshots:   0 total
Time:        11.939s
Ran all test suites matching /test/i.
Done in 13.07s.

real	0m13.345s
user	0m24.808s
sys	0m2.792s
```

**Comment by peterbe** [comment]
See the latest commit ^. I picked 2 because that's what someone suggest here: https://discuss.circleci.com/t/memory-problems-with-jest-and-workers/10297/4

Shame there's no way to document this in `package.json`.

**Comment by peterbe** [comment]
Wonderful! Now the test block only takes 20 seconds.

**Comment by mythmon** [comment]
I imagine that CI is running on a docker-host that really does have 36 physical cores, but we are limited by a CPU quota. That's pretty normal for Docker. It's too bad that Jest didn't pick up on this automatically. Maybe we should file a bug for that? And even if we did have access to all 36 cores, we clearly don't have the memory to support that. It's probably less reasonable to ask Jest to autodetect that for us.

Nice work spotting that we were running with too many parallel jobs. I don't think I would have thought to check that.

File: src/workflows/recipes/pages/RecipeListingPage.js (hunk 1 above)
Comment (rehandalal) [reply src/workflows/recipes/pages/RecipeListingPage.js]: it could be that too.

**Pull request merged**